src/rhsmlib/cloud/README.md

```diff
@@ -1,8 +1,189 @@
+
+    def __init__(self, hw_info):
+        """
+        Initialize instance of FooCloudDetector
+        """
+        super(FooCloudProvider, self).__init__(hw_info)
+
+    def is_vm(self):
+        """
+        Is system running on virtual machine or not
+        :return: True, when machine is running on VM; otherwise return False
+        """
+        return super(FooCloudProvider, self).is_vm()
+
+    def is_running_on_cloud(self):
+        """
+        Try to guess if cloud provider is Foo using collected hardware information (output of dmidecode,
+        virt-what, etc.)
+        :return: True, when we detected sign of Foo in hardware information; Otherwise return False
+        """
+
+        # TODO: Check if this is true for Foo Cloud Provider
+        if self.is_vm() is False:
+            return False
+
+        # This is valid for virtual machines running on Foo
+        if 'dmi.chassis.asset_tag' in self.hw_info and \
+                self.hw_info['dmi.chassis.asset_tag'] == 'FOO':
+            return True
+        # In other cases return False
+        return False
+
+    def is_likely_running_on_cloud(self):
+        """
+        Return non-zero value, when the machine is virtual machine and it is running on Foo
+        hypervisor and some Foo string can be found in output of dmidecode
+        :return: Float value representing probability that vm is running on Foo
+        """
+        probability = 0.0
+
+        # TODO: Check if this is true for Foo Cloud Provider
+        if self.is_vm() is False:
+            return 0.0
+
+        if 'dmi.chassis.asset_tag' in self.hw_info and \
+                self.hw_info['dmi.chassis.asset_tag'] == 'FOO':
+            probability += 0.3
+
+        # Try to find "Foo" keyword in output of dmidecode
+        found_foo = False
+        for hw_item in self.hw_info.values():
+            if type(hw_item) != str:
+                continue
+            if 'foo' in hw_item.lower():
+                found_foo = True
+        if found_foo is True:
+            probability += 0.3
+
+        return probability
+
+
+    def _get_metadata_from_cache(self) -> Union[str, None]:
+        """
+        TODO: implement something or return None
+        """
+        raise NotImplementedError
+
+    def _get_data_from_server(self, data_type, url):
+        """
+        This method tries to get data from server using GET method
+        :param data_type: string representation of data type used in log messages (e.g. "metadata", "signature")
+        :param url: URL of GET request
+        :return: String of body, when request was successful; otherwise return None
+        """
+        return super(FooCloudProvider, self)._get_data_from_server(data_type, url)
+
+    def _get_metadata_from_server(self) -> Union[str, None]:
+        """
+        Try to get metadata from server
+        :return: String with metadata or None
+        """
+        return super(FooCloudProvider, self)._get_metadata_from_server()
```

Comment:
these seems not needed to reimplement in subclasses

Reply:
This is example describing complete API. `AWSCloudProvider` reimplemented this method.